### PR TITLE
Custom kernel priors

### PR DESCRIPTION
**Pull request by AVHopp**
This PR introduces custom priors for kernels. It provides an ABC `Prior` as well as a `GammaPrior` class and corresponding tests.

Note that this PR does not yet introduce custom `ScaleKernels` and only adjusts the priors at the point where they are used for our custom `MaternKernels`. At other parts of the code, we technically still use the original `GPyTorch` priors, which is a bit hidden by the `.to_gyptorch()` call. This will all be fixed in upcoming PRs.

Further note that the kernel serialization currently fails. @AdrianSosic and I already briefly investigated the error and did not manage to track it down, and agreed to investigate this issue once the code is pushed (which is the case now)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -12,7 +12,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Hypothesis strategies for acquisition functions
 - `Kernel` base class allowing to specify kernels
 - `MaternKernel` class can be chosen for GP surrogates
-- `hypothesis` strategies and roundtrip test for kernels, constraints and objectives
+- `GammaPriors` class can be chosen for `MaternKernel`s
+- `hypothesis` strategies and roundtrip test for kernels, constraints, objectives and priors
 
 ### Changed
```
Comment (AVHopp) [inline CHANGELOG.md:18]: **NOTE** There seems to be an unrelated error in the doc building due to some stuff that seems to have changed in other packages. I am currently investigating that and will resolve here once this was fixed.

File: tests/hypothesis_strategies/priors.py [hunk 2]
```diff
@@ -0,0 +1,11 @@
+"""Hypothesis strategies for priors."""
```
Comment (AVHopp) [inline tests/hypothesis_strategies/priors.py:1]: @AdrianSosic here is the error that we discussed earlier, screenshot taken from the logs: 
image

**Review by Scienfitz — COMMENTED** [review]
first round of quick comments

File: baybe/kernels/priors/gamma.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
+"""A Gamma prior parameterized by concentration and rate."""
```
[the remark this answers is not in the document]
Comment (AVHopp) [reply baybe/kernels/priors/gamma.py:1]: We could it similar to how we did it wit the kernels where we agreed to have one `base` fiel for the base class and one `basic` file for "basic" kernels. I'll adjust it accordingly.

File: baybe/kernels/priors/gamma.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+from attrs.validators import gt
+
+from baybe.kernels.priors.base import Prior
+
+
+@define
+class GammaPrior(Prior):
+    """A Gamma prior parameterized by concentration and rate."""
+
+    concentration: float = field(validator=gt(0.0))
```
[the remark this answers is not in the document]
Comment (AVHopp) [reply baybe/kernels/priors/gamma.py:13]: I was not sure which values to choose here as defaults since we basically choose between 4 different sets of default values ourselves currently. So which should be chosen?

File: baybe/surrogates/gaussian_process.py [hunk 5]
```diff
@@ -104,12 +107,15 @@ def _fit(self, searchspace: SearchSpace, train_x: Tensor, train_y: Tensor) -> No
         # create GP mean
         mean_module = gpytorch.means.ConstantMean(batch_shape=batch_shape)
 
+        # If no kernel is provided, we construct one from our priors
+        if self.kernel is None:
+            self.kernel = MaternKernel(lengthscale_prior=lengthscale_prior[0])
```
[the remark this answers is not in the document]
Comment (AVHopp) [reply baybe/surrogates/gaussian_process.py:112]: Not sure since this happens in the middle of `_fit` since this is the place where we choose the priors. Can we still use `attrs` then?

File: baybe/kernels/priors/gamma.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (AVHopp) [reply baybe/kernels/priors/gamma.py]: Done.

File: tests/hypothesis_strategies/priors.py (hunk 2 above)
Comment (AdrianSosic) [reply tests/hypothesis_strategies/priors.py:1]: Problem identified, as mentioned via chat. `kw_only=True` missing in the `lengthscale_prior` field of the `Kernel` base class

File: CHANGELOG.md [hunk 6]
```diff
@@ -12,7 +12,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Hypothesis strategies for acquisition functions
 - `Kernel` base class allowing to specify kernels
 - `MaternKernel` class can be chosen for GP surrogates
-- `hypothesis` strategies and roundtrip test for kernels, constraints and objectives
+- `GammaPrior` class can be chosen for `MaternKernel`s
```
Comment (AdrianSosic) [inline CHANGELOG.md:15]: Not quite accurate:
* It can be chosen for ALL kernels, not just the MaternKernel
* and it's not chosen as Kernel choice but for the lenghtscale of the kernel

File: baybe/kernels/priors/gamma.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Scienfitz) [reply baybe/kernels/priors/gamma.py]: i see, I was assuming the Gpytorch kernels have defaults, but it seems they dont, guess then we dont have to bother with them although I always find it nice
isnt there a "general" setting of these settings that just makes an extremely broad prior?

File: baybe/kernels/base.py [hunk 7]
```diff
@@ -17,12 +19,20 @@
 class Kernel(ABC, SerialMixin):
     """Abstract base class for all kernels."""
 
+    lengthscale_prior: Optional[Prior] = field(default=None, kw_only=True)
+    """An optional prior that can be applied to the lengthscale."""
```
Comment (AdrianSosic) [inline baybe/kernels/base.py:23]: I think "apply" sounds a bit weird here
```suggestion
    """An optional prior on the kernel lengthscale."""
```

File: baybe/kernels/base.py [hunk 8]
```diff
@@ -17,12 +19,20 @@
...
+
     def to_gpytorch(self, *args, **kwargs):
         """Create the gpytorch representation of the kernel."""
         import gpytorch.kernels
 
         kernel_cls = getattr(gpytorch.kernels, self.__class__.__name__)
         fields_dict = filter_attributes(object=self, callable_=kernel_cls.__init__)
+
+        # If a lengthscale_prior was chosen, we manually add it to the dictionary
```
Comment (AdrianSosic) [inline baybe/kernels/base.py:32]:
```suggestion
        # If a lengthscale prior was chosen, we manually add it to the dictionary
```

File: baybe/kernels/priors/basic.py [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+from baybe.kernels.priors.base import Prior
+
+
+@define
+class GammaPrior(Prior):
+    """A Gamma prior parameterized by concentration and rate."""
+
+    concentration: float = field(validator=gt(0.0))
+    """The concentration."""
+    rate: float = field(validator=gt(0.0))
```
Comment (AdrianSosic) [inline baybe/kernels/priors/basic.py:15]: blank line in between

File: baybe/kernels/priors/basic.py [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+from attrs.validators import gt
+
+from baybe.kernels.priors.base import Prior
+
+
+@define
+class GammaPrior(Prior):
+    """A Gamma prior parameterized by concentration and rate."""
+
+    concentration: float = field(validator=gt(0.0))
```
Comment (AdrianSosic) [inline baybe/kernels/priors/basic.py:13]:
```suggestion
    concentration: float = field(converter=float, validator=gt(0.0))
```

File: baybe/kernels/priors/basic.py (hunk 9 above)
Comment (AdrianSosic) [inline baybe/kernels/priors/basic.py:15]:
```suggestion
    rate: float = field(converter=float, validator=gt(0.0))
```

File: baybe/kernels/priors/basic.py [hunk 11]
```diff
@@ -0,0 +1,16 @@
+"""A Gamma prior parameterized by concentration and rate."""
```
Comment (AdrianSosic) [inline baybe/kernels/priors/basic.py:1]: Module docstring should be generic, not just about gamma priors

File: baybe/surrogates/gaussian_process.py [hunk 12]
```diff
@@ -46,7 +46,8 @@ def _fit(self, searchspace: SearchSpace, train_x: Tensor, train_y: Tensor) -> No
         import botorch
         import gpytorch
         import torch
-        from gpytorch.priors import GammaPrior
+
+        from baybe.kernels.priors import GammaPrior
```
Comment (AdrianSosic) [inline baybe/surrogates/gaussian_process.py:50]: move top level

File: baybe/surrogates/gaussian_process.py [hunk 13]
```diff
@@ -27,7 +27,7 @@ class GaussianProcessSurrogate(Surrogate):
     # See base class.
 
     # Object variables
-    kernel: Kernel = field(factory=MaternKernel)
+    kernel: Optional[Kernel] = field(default=None)
```
Comment (AdrianSosic) [inline baybe/surrogates/gaussian_process.py:30]: Normally I would not accept having the field optional here, but I understand that this is intertwined with the GP refactoring, so I'm fine with having it as a temporary workaround. But this will be changed back later

File: tests/hypothesis_strategies/priors.py
Comment (AdrianSosic) [inline tests/hypothesis_strategies/priors.py:1]: Can you also extend the kernel strategy such that it includes the new priors? For this, please first introduce a `priors` strategy that you can then use inside the kernel, augmented such that also `None` is generated at that level

**Review by AdrianSosic — COMMENTED** [review]
Hi @AVHopp, here my current feedback. Thanks for drafting 👍🏼

File: CHANGELOG.md (hunk 6 above)
Comment (AVHopp) [reply CHANGELOG.md:15]: So what would you suggest? "`GammaPrior` can now be chosen as lengthscale prior"?

File: baybe/kernels/base.py (hunk 7 above)
Comment (AVHopp) [reply baybe/kernels/base.py:23]: I think I took that from the gpytorch formulation, will adjust :)

File: baybe/surrogates/gaussian_process.py (hunk 13 above)
Comment (AVHopp) [reply baybe/surrogates/gaussian_process.py:30]: Can you elaborate on that? I guess what you mean is normally you would not make this optional but use a factory which is currently not possible due to the fact that the priors are currently not chosen during initialization, or what do you mean?

File: baybe/kernels/priors/gamma.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/kernels/priors/gamma.py]: (and of course, we can always add defaults later on if we think it's reasonable)

File: baybe/surrogates/gaussian_process.py (hunk 13 above)
Comment (AdrianSosic) [reply baybe/surrogates/gaussian_process.py:30]: yes exactly, just answered [above](https://github.com/emdgroup/baybe/pull/219#discussion_r1579616590) 😄

File: baybe/surrogates/gaussian_process.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (AVHopp) [reply baybe/surrogates/gaussian_process.py:112]: So this means there won't be a change here and I can resolve?

File: baybe/surrogates/gaussian_process.py (hunk 13 above)
Comment (AdrianSosic) [reply baybe/surrogates/gaussian_process.py:30]: Ok, if you don't like the previous, [this](https://github.com/emdgroup/baybe/pull/219#discussion_r1579695981) will drive you crazy

File: baybe/surrogates/gaussian_process.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/surrogates/gaussian_process.py:112]: From my point yes. Will do so. @Scienfitz: please unresolve if necessary

File: baybe/kernels/priors/gamma.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/kernels/priors/gamma.py]: I see it as part of the GP refactoring PR that we assigned to me because you basically can not solve it here. So will close the thread. @Scienfitz: please unresolve if necessary

File: baybe/kernels/base.py (hunk 8 above)
Comment (AVHopp) [reply baybe/kernels/base.py]: Changed.

File: baybe/kernels/priors/basic.py (hunk 11 above)
Comment (AVHopp) [reply baybe/kernels/priors/basic.py]: Changed, please resolve if happy.

File: CHANGELOG.md (hunk 6 above)
Comment (AVHopp) [reply CHANGELOG.md:15]: Changed it accordingly, please un-resolve if un-happy

File: baybe/surrogates/gaussian_process.py (hunk 12 above)
Comment (AVHopp) [reply baybe/surrogates/gaussian_process.py]: Done.

File: tests/hypothesis_strategies/priors.py
Comment (AVHopp) [reply tests/hypothesis_strategies/priors.py:1]: Made an attempt of including this, is this how you imagined it to be?

Comment (AdrianSosic) [reply tests/hypothesis_strategies/priors.py:1]: Have pushed two additional commits (feel free to squash in case you rebase stuff):
* Changed order of strategies (FYI: for shrinking, simple strategies should always come first)
* Included the generic prior strategy

You can resolve if no further questions

**Review by AdrianSosic — COMMENTED** [review]
@AVHopp: I think the PR is now ready. Could you do a quick rebase to squash all the little fixup commits and then we can merge (unless @Scienfitz has more comments, of course) 🚀

File: baybe/kernels/priors/basic.py [hunk 14]
```diff
@@ -0,0 +1,16 @@
+"""Priors that can be used for kernels."""
+from attrs import define, field
+from attrs.validators import gt
+
+from baybe.kernels.priors.base import Prior
+
+
+@define
```
Comment (Scienfitz) [inline baybe/kernels/priors/basic.py:8]: should we consider frozen and or slotted?

File: baybe/surrogates/gaussian_process.py [hunk 15]
```diff
@@ -72,29 +72,31 @@ def _fit(self, searchspace: SearchSpace, train_x: Tensor, train_y: Tensor) -> No
...
         )
 
+        # TODO Until now, only the kernels use our custom priors, hence the explicit
+        # to_gpytorch() calls for all others
         # low D priors
         if train_x.shape[-1] < 10:
             lengthscale_prior = [GammaPrior(1.2, 1.1), 0.2]
-            outputscale_prior = [GammaPrior(5.0, 0.5), 8.0]
-            noise_prior = [GammaPrior(1.05, 0.5), 0.1]
+            outputscale_prior = [GammaPrior(5.0, 0.5).to_gpytorch(), 8.0]
```
Comment (Scienfitz) [inline baybe/surrogates/gaussian_process.py:80]: wouldnt it make sense to shift the `to_gpytorch` calls to the code below where they are actually passed to a gpytorch function. This here feels very messy due to the mixing

File: tests/serialization/test_prior_serialization.py [hunk 16]
```diff
@@ -0,0 +1,13 @@
+"""Test serialization of kernels."""
+
+from hypothesis import given
+
+from baybe.kernels.priors import GammaPrior
+from tests.hypothesis_strategies.priors import gamma_priors
+
+
+@given(gamma_priors)
+def test_gamma_prior_kernel_roundtrip(prior: GammaPrior):
```
Comment (Scienfitz) [inline tests/serialization/test_prior_serialization.py:10]: i know we only have gammapriors at the moment

but given youve already introduce a generic `priors` strategy why not make this test name and hypothesis general too? it would mean any update and addition of enw priors wouldnt require any change here anymore

**Review by Scienfitz — APPROVED** [review]

File: baybe/kernels/priors/basic.py (hunk 14 above)
Comment (AdrianSosic) [reply baybe/kernels/priors/basic.py:8]: Oh yes, absolutely! Both for the kernels and the priors it's missing!

File: tests/serialization/test_prior_serialization.py (hunk 16 above)
Comment (AdrianSosic) [reply tests/serialization/test_prior_serialization.py:10]: Upvote! This was exactly the motivation for the generic `priors` strategy: to avoid later changes in other places 👍🏼

File: CHANGELOG.md (hunk 1 above)
Comment (AVHopp) [reply CHANGELOG.md:18]: Will be fixed by another PR

File: tests/serialization/test_prior_serialization.py (hunk 16 above)
Comment (AVHopp) [reply tests/serialization/test_prior_serialization.py]: Changed. 

BTW: You realized stuff like this, but NOBODY saw that the module docstring said `"""Test serialization of kernels."""` :D

File: baybe/surrogates/gaussian_process.py (hunk 15 above)
Comment (AVHopp) [reply baybe/surrogates/gaussian_process.py:80]: I did that on purpose to make clear that we still technically require these objects to be `gpytorch` objects. But I will simply move these calls to the places where the conversion needs to happen and then resolve.

File: baybe/kernels/priors/basic.py (hunk 14 above)
Comment (AVHopp) [reply baybe/kernels/priors/basic.py:8]: No: The base classes are frozen (this is explicitly declared), and slots is true by default. Also, according to the attrs documentation, the child classes are thus also frozen (see point 5):
image
So it's not missing?

Comment (AdrianSosic) [reply baybe/kernels/priors/basic.py:8]: Yes, the slots is fine, I know. But thanks for the info on inherited frozenness (<- is that a word 🤔), wasn't aware of that. But still, I think at this point we should add it for consistency reasons, just because we have it added it explicitly for ALL OTHER subclasses in the codebase. So in principle we can omit it, but if we do so, this should rather happen in one of the general cleanup PRs in the future (and it's not really high priority). So would favor consistency for now ...

Comment (AVHopp) [reply baybe/kernels/priors/basic.py:8]: Thy wish shall be granted.

Comment (AdrianSosic) [reply baybe/kernels/priors/basic.py]: Thanks. In fact, it has another advantage: in case we ever need to remove frozenness for the base class, we still have it for the subclasses and can control it case by case. But that's rather minor. Thanks for changing ⭐

File: tests/serialization/test_prior_serialization.py (hunk 16 above)
Comment (AdrianSosic) [reply tests/serialization/test_prior_serialization.py]: Was supposed to be a test ... you passed it, congratulations 🤓

File: CHANGELOG.md [hunk 17]
```diff
@@ -12,7 +12,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Hypothesis strategies for acquisition functions
 - `Kernel` base class allowing to specify kernels
 - `MaternKernel` class can be chosen for GP surrogates
-- `hypothesis` strategies and roundtrip test for kernels, constraints and objectives
+- `GammaPrior` can now be chosen as lengthscale prior
```
Comment (AVHopp) [inline CHANGELOG.md:15]: @AdrianSosic : Just a ping that rebasing is done. Note that I was quite radical in trimming down the commits, but I have a backup in hand in case you'd prefer more commits.

File: tests/serialization/test_prior_serialization.py (hunk 16 above)
Comment (Scienfitz) [reply tests/serialization/test_prior_serialization.py]: i dont know why, but I always ignore titles of modules and powerpoint slides, they are simply irrelevant to me  🤷‍♂️

File: CHANGELOG.md (hunk 17 above)
Comment (AVHopp) [reply CHANGELOG.md:15]: Actually disagree here. If you already see an issue here, why wait? I also already pushed a draft version of the scale kernel approach, so please have a look there, because that is (besides the fact that it is not yet ready to merge due to this not yet being merged) already ready for review (put still a bit messy due to it being based on this)
